src/librustc_interface/queries.rs

```diff
@@ -278,35 +276,54 @@ impl<'comp> Queries<'comp> {
         })
     }
 
-    pub fn link(&self) -> Result<&Query<()>> {
-        self.link.compute(|| {
-            let sess = self.session();
+    pub fn linker(self) -> Result<Linker> {
+        let dep_graph = self.dep_graph()?;
+        let prepare_outputs = self.prepare_outputs()?;
+        let ongoing_codegen = self.ongoing_codegen()?;
 
-            let ongoing_codegen = self.ongoing_codegen()?.take();
+        let sess = self.session().clone();
+        let codegen_backend = self.codegen_backend().clone();
 
-            self.codegen_backend().join_codegen_and_link(
-                ongoing_codegen,
-                sess,
-                &*self.dep_graph()?.peek(),
-                &*self.prepare_outputs()?.peek(),
-            ).map_err(|_| ErrorReported)?;
-
-            Ok(())
+        Ok(Linker {
+            sess,
+            dep_graph: dep_graph.take(),
+            prepare_outputs: prepare_outputs.take(),
+            ongoing_codegen: ongoing_codegen.take(),
+            codegen_backend,
         })
     }
 }
 
+pub struct Linker {
+    sess: Lrc<Session>,
+    dep_graph: DepGraph,
+    prepare_outputs: OutputFilenames,
+    ongoing_codegen: Box<dyn Any>,
+    codegen_backend: Lrc<Box<dyn CodegenBackend>>,
+}
+
+impl Linker {
+    pub fn link(self) -> Result<()> {
+        self.codegen_backend.join_codegen_and_link(
+            self.ongoing_codegen,
+            &self.sess,
+            &self.dep_graph,
+            &self.prepare_outputs,
+        ).map_err(|_| ErrorReported)
+    }
+}
+
 impl Compiler {
     // This method is different to all the other methods in `Compiler` because
     // it lacks a `Queries` entry. It's also not currently used. It does serve
     // as an example of how `Compiler` can be used, with additional steps added
     // between some passes. And see `rustc_driver::run_compiler` for a more
     // complex example.
     pub fn enter<'c, F, T>(&'c self, f: F) -> Result<T>
-        where F: for<'q> FnOnce(&'q Queries<'c>) -> Result<T>
+        where F: FnOnce(Queries<'c>) -> Result<T>
```

Comment:
This closure should take `&Queries<'_>`. We don't want to give out ownership. The `'c` lifetime also isn't needed here. The comment above is also for the wrong function.